plugins/BEdita/Core/src/Model/Table/ObjectTypesTable.php

```diff
@@ -196,16 +197,25 @@ public function get($primaryKey, $options = [])
 
     /**
      * Set default parent on creation if missing.
+     * Prevent delete if type is abstract and a subtype exists.
+     *
+     * Controls are performed here insted of `beforeSave()` or `beforeDelete()`
+     * in order to be executed before corresponding methods in `TreeBehavior`.
      *
      * @param \Cake\Event\Event $event The event dispatched
      * @param \Cake\Datasource\EntityInterface $entity The entity to save
-     * @return void
+     * @throws \Cake\Network\Exception\ForbiddenException if operation on entity is not allowed
      */
     public function beforeRules(Event $event, EntityInterface $entity)
     {
         if ($entity->isNew() && empty($entity->parent_id)) {
             $entity->parent_id = self::DEFAULT_PARENT_ID;
         }
+        $eventData = $event->getData();
+        if (!empty($eventData['operation']) && $eventData['operation'] === 'delete'
```

Comment:
I miss the point. Why isn't this logic in `beforeDelete()`?

Reply:
Ok… so the point is that performing a query `SELECT ... FROM object_types WHERE parent_id = 123` in `ObjectTypesTable::beforeDelete()` always returns an empty set because the relevant nodes have already been deleted by `TreeBehavior::beforeDelete()`, right?

I'd suggest to apply these changes instead:
```patch
--- plugins/BEdita/Core/src/Model/Table/ObjectTypesTable.php	(revision 5a0b7fab88bc12e610c0362eba2a8091da86fa06)
+++ plugins/BEdita/Core/src/Model/Table/ObjectTypesTable.php	(revision )
@@ -43,6 +43,8 @@
  * @method \BEdita\Core\Model\Entity\ObjectType[] patchEntities($entities, array $data, array $options = [])
  * @method \BEdita\Core\Model\Entity\ObjectType findOrCreate($search, callable $callback = null, $options = [])
  *
+ * @mixin \Cake\ORM\Behavior\TreeBehavior
+ *
  * @since 4.0.0
  */
 class ObjectTypesTable extends Table
@@ -209,12 +211,7 @@
     public function beforeRules(Event $event, EntityInterface $entity)
     {
         if ($entity->isNew() && empty($entity->parent_id)) {
+            $entity->set('parent_id', self::DEFAULT_PARENT_ID);
-            $entity->parent_id = self::DEFAULT_PARENT_ID;
-        }
-        $eventData = $event->getData();
-        if (!empty($eventData['operation']) && $eventData['operation'] === 'delete'
-            && $entity->get('is_abstract') && $this->find()->where(['parent_id' => $entity->id])->first()) {
-            throw new ForbiddenException(__d('bedita', 'Abstract type with existing subtypes'));
         }
     }
 
@@ -239,6 +236,9 @@
      */
     public function beforeDelete(Event $event, EntityInterface $entity)
     {
+        if ($entity->get('is_abstract') && $this->childCount($entity) > 0) {
+            throw new ForbiddenException(__d('bedita', 'Abstract type with existing subtypes'));
+        }
         if (TableRegistry::get('Objects')->exists(['object_type_id' => $entity->id])) {
             throw new ForbiddenException(__d('bedita', 'Objects of this type exist'));
         }
```

This way we rely on `TreeBehavior::childCount($entity)` to know how many descendants the current node has. This _is not_ achieved through a query unless explicitly asked passing a second argument as `true`, but rather uses NSM left and right indexes to know how many nodes are between those values — basically `(right - left - 1) / 2`.

Raising an exception should hopefully roll back the transaction, so data integrity should be safe. 😱

Reply:
This will work, but honestly I'd prefer not to rely on a rollback for these kind of checks.
But using `childCount()` instead of `find()->first()` is better...

Reply:
Yes, you're right.

Also, in that case a lot of rows would be deleted by tables like `objects`, `object_relations`, `object_properties`, … due to foreign key constraints, so we would drop tons of rows and _then_ trigger a rollback. We better avoid.